### PR TITLE
Experimental SQLite support BT-54

### PR DESCRIPTION
**Pull request by kshakir**
- Upgraded Liquibase to latest
- ~Workaround Liquibase UniqueConstraint "caching" bug~ EDIT: Bug was fixed in liquibase!
- Fixed S3 SPI config to avoid Liquibase warnings
- Removed unused DB upgrade environment variables
- Test various DB combinations using centaur local

### CODE REVIEW

**Comment by aednichols** [comment]
TOL:

It seems that SQLite is a superset of what is supported by HSQL. It is also very broadly popular.

Could we _replace_ HSQL with SQLite so that we can aren't maintaining two embedded database implementations?

**Comment by kshakir** [comment]
> Could we replace HSQL with SQLite so that we can aren't maintaining two embedded database implementations?

Created BT-56 to track the issue and steps to getting there. TL;DR we don't CI test the embedded DB with real WDLs so I'd be hesitant to change to SQLite until we can know before users if/when we break it.

File: docs/Configuring.md
Comment (rhpvorderman) [inline docs/Configuring.md:423]: The long connection timeout is specific to the HSQLDB file database though. 3GB (!) of files is not uncommon, and that is after a single workflow run with a few hundred jobs. I wonder how big the sqlite database will end up to be.

**Comment by rhpvorderman** [comment]
Wow! Thanks @kshakir! This is great news! Thanks for putting your effort in this! 
I think I will start a few pipelines to see the effect on memory usage and storage space and compare that with HSQLdb.

@aednichols 
> It seems that SQLite is a superset of what is supported by HSQL. It is also very broadly popular.

> Could we replace HSQL with SQLite so that we can aren't maintaining two embedded database implementations?

Replacing HSQL with SQLite and making SQLite default has multiple advantages:
- HSQL uses a lot of memory for the in-memory database. SQLite is probably more efficient (testing required)
- With SQLite a file-based database could be enabled by default (in `<<CROMWELL_ROOT>>/cromwell-db.sqlite`) which would enable call-caching by default for people using it from the command line with the run command.
- If the filebased database is enabled by default, there is no need for most users to get into the details of getting a proper database server running.

However HSQL has the advantage that it is very well supported by slick and other java/scala practice (HSQL is itself written in Java) and that it requires minimal code to get it working. By contrast SQLite needs more code as shown in this PR. It is thus more costly to support SQLite.

**Comment by rhpvorderman** [comment]
Okay preliminary testing. On a workflow that creates multiple jobs at once I get these errors: `Failure writing to call cache: [SQLITE_BUSY] the database file is locked (database is locked)`

This has to do probably with SQLITE not supporting multiple threads or something similar. It is probably reproducible by creating a mock scatter workflow that spawns a 1000 jobs at once.

File: docs/Configuring.md
Comment (rhpvorderman) [inline docs/Configuring.md:379]: I just did some testing and found that when the directory does not exist, cromwell will hang. 
A working URL would be
`url = "jdbc:sqlite:file:cromwell-db.sqlite?foreign_keys=true&date_class=text"`

**Comment by rhpvorderman** [comment]
I did some further debugging and I found that cromwell creates a SlickDatabase object twice. Once for the database and once for the metadata. That is probably were the conflict comes from.

Using the following configuration the problem could not be reproduced anymore. I am now trying it on a production workflow on our cluster.

```hocon
database {
  profile = "slick.jdbc.SQLiteProfile$"
  db {
    driver = "org.sqlite.JDBC"
    url = "jdbc:sqlite:cromwell.sqlite?foreign_keys=true&date_class=text"
    numThreads=1
  }
  metadata {
   profile = "slick.jdbc.SQLiteProfile$"
    db {
      driver = "org.sqlite.JDBC"
      url = "jdbc:sqlite:cromwell-metadata.sqlite?foreign_keys=true&date_class=text"
      numThreads=1
    }
  }
}

```

I am currently looking in how to make the metadata and engine use the same connection when they are using the same configuration.

**Comment by rhpvorderman** [comment]
I did some benchmarking over the weekend. I ran a 1000+ job workflow on both the HSQL database with overflow file, and with sqlite.

The Sqlite database creates 99M in files. The HSQLDB creates 9.5G in files, that is 100 times more... I restarted the workflow to see if the call caching worked properly. With the HSQLDB there was no issue in restarting. With SQLite everything worked fine until at some point early in the workflow cromwell hung. I interrupted the process, and cromwell started to shut down gracefully. `WriteMetadataActor shutting down: processing 108720 queued messages`. The processing of these messages takes more than half an hour.  

The problem here is twofold: the SQLite backend is significantly slower than the HSQL in-memory database with overflow file (as expected) and the enormous amount of messages that Cromwell produces totally swamps it. (Judging from the shutdown scroll the rate is approximately 40 messages per second or 25ms per message processing time, sqlite should be able to work faster than that).

**Comment by kshakir** [comment]
Hi @rhpvorderman-

Thanks for giving this PR a full work out. I don't have a lot of experience running SQLite so for others who might want to follow in your steps your experience is very valuable.

A few requests/questions:
- For the other reviewers do you have an itemized list of things this PR would need before it should be merged?
- Based on your testing is anything missing from our Centaur regression test config?
- Do you have the time to submit changes to the PR to polish it up (hopefully mostly docs)?

For the lurkers, I'll try to reply with my comments and hypotheses in a follow-up reply this week. Ex: when/why did we create two databases in Cromwell, and how I think that affects SQLite which only wants \*one\* connection per DB.

**Comment by rhpvorderman** [comment]
@kshakir Thanks for all your efforts in getting this PR working. Upon more careful inspection I saw that you already had found the metadata/engine should be separate issue. Sorry for double reporting.

* The regression testing looks OK to me :+1:
* I have rearranged the docs a bit. SQLite is suggested first and contrasted with other databases. HSQLDB is listed after that to make users aware of the option, but it is also made clearly that this is for very specific use cases. I updated the hsql and sqlite config examples a bit.
* For the liquibase spec testing I ensured that an actual file database is used.

I did some testing with the in-memory database for only metadata, but that failed for some reason when running it on a big pipeline. At least the file-based database is working properly, which is also what we test in all the tests.

File: CHANGELOG.md
Comment (grsterin) [inline CHANGELOG.md:11]: "Unsupported support" sounds weird, so I'd leave only "experimental" part here or somehow rephrase this.

**Comment by grsterin** [comment]
One comment regarding wording in changelog. Otherwise looks good.

**Review by grsterin — APPROVED** [review]

**Review by aednichols — CHANGES_REQUESTED** [review]
I'd like to hold this PR until the new year so we can have a team discussion.

We've previously talked about how we want to _remove_ support for certain seldom-used databases to reduce maintenance burden, and everyone seemed in favor.

This change has implications for product strategy and we should discuss it more fully than two thumbs.

**Comment by aednichols** [comment]
@rhpvorderman we are not able to muster the resources to bring this across the finish line right now.

It is definitely something we will keep an eye on in the future if we have more funding for supporting community-oriented features.

**Comment by rhpvorderman** [comment]
@aednichols. No problem. It also turns out that the SQLite file database performs rather slowly. This is due to cromwell doing a lot of transactions, and these are severely bottlenecked by filesystem i/o limits. The amount of i/o operations per second on NFS is rather low so it chokes cromwell performance quite a bit when running a job with multiple samples. This shows that there is room for improvement in Cromwell's design (why are there more than a hundred thousand database interactions for a 2000 job run?) but that is another matter entirely.
HSQLDB with overflow file has problems of its own, but it is more performant (once the 10 GB+ file has been read at least).

This branch works, so if the need arises I can maintain a separate release of cromwell with these changes patched in. I have done that before for some changes in dev that could not wait.

For now much much thanks for all your support in getting it this far :heart:.

**Comment by aednichols** [comment]
@rhpvorderman thanks for understanding.

Cromwell is basically designed to run with a dedicated database instance and as such does not attempt to economize on database transactions.

All of our internal customers and most external ones do use a dedicated instance. We definitely recognize there are other use cases that we support subject as best we can, albeit far from perfectly.

**Pull request closed without merging**
